### PR TITLE
Feature/label on profile

### PR DESCRIPTION
**Pull request by eileenabadi**
Missing:
-. Connection to add label directly from Profile page

FE:
-. Create label tag on profile
-. Create add label page
-. Create add label on profile

BE:
-. Create table Label
-. Create connection with FE

### CODE REVIEW

File: client/src/app/main/profile/edit/profile-edit.component.ts [hunk 1]
```diff
@@ -32,37 +37,68 @@ export class ProfileEditComponent implements OnInit {
...
+  public onDeselectAll() {
+    this.selectedItems = [];
   }
 
   public saveProfile() {
     this._profileService.update(this.profile.id, this.editProfileFormGroup.value)
-      .subscribe((updatedProfile: Profile) => {
-          this.profile = updatedProfile;
+      .subscribe((updateProfile: Profile) => {
+          this.profile = updateProfile;
```
Comment (jpastena) [inline client/src/app/main/profile/edit/profile-edit.component.ts]: update**d**Profile seems fine

File: client/src/app/main/profile/label/profile-label.component.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+              private _alertService: AlertService,
+              private _formBuilder: FormBuilder) { }
+
+  public ngOnInit() {
+    this.getAllLabel();
+    this.addLabelForm = this._formBuilder.group(NewLabelFormFields);
+    this.addLabelForm.reset();
+  }
+
+  public getAllLabel() {
```
Comment (jpastena) [inline client/src/app/main/profile/label/profile-label.component.ts]: getAllLabes*  
in the other components I usually called this method `retrieveLabels`

File: client/src/app/main/profile/label/profile-label.component.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    this._labelService.find()
+      .subscribe((labels: LabelInterface[]) => {
+          this.labels = labels;
+          console.log(labels);
```
Comment (jpastena) [inline client/src/app/main/profile/label/profile-label.component.ts]: remove

File: client/src/app/shared/services/label/label.service.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import { Observable } from 'rxjs';
+import {LabelInterface} from "../../interfaces/label.interface";
+import {Profile} from "../../models/profile.models";
+
+@Injectable({
+  providedIn: 'root'
+})
+export class LabelService {
+  public actionUrl: string;
+  public dropdownSettings = {
```
Comment (jpastena) [inline client/src/app/shared/services/label/label.service.ts]: - is a Type Available for this object? (it depends on the npm package that you used)
- not sure if this information should be located in the Service (a service should be only accountable to the backend connection, while here is used for "html" stuff)

File: client/src/app/shared/services/label/label.service.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+              private _configuration: Configuration) {
+    this.actionUrl = `${_configuration.apiHost}${_configuration.apiPrefix}label`;
+  }
+
+  public find(): Observable<Object> {
+    return this._httpClient.get(this.actionUrl);
+  }
+
+  public create(label: LabelInterface) {
+    return this._httpClient.post(`${this.actionUrl}`, label);
```
Comment (jpastena) [inline client/src/app/shared/services/label/label.service.ts]: the ``${}`` is not necessary here, see the get method above

File: client/src/app/shared/services/label/label.service.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    itemsShowLimit: 10,
+    allowSearchFilter: true
+  };
+
+  constructor(private _httpClient: HttpClient,
+              private _configuration: Configuration) {
+    this.actionUrl = `${_configuration.apiHost}${_configuration.apiPrefix}label`;
+  }
+
+  public find(): Observable<Object> {
```
Comment (jpastena) [inline client/src/app/shared/services/label/label.service.ts]: avoid `Object` and `any` as much as possible.  
here you can return `Observable<LabelInterface>`, but you are to specify that on the httpclient as well.  
use `this._httpClient.get<LabelInterface>(...)` to make it work

Comment (jpastena) [reply client/src/app/shared/services/label/label.service.ts]: Apply the same changes to the other methods

File: database/createDB.sql [hunk 7]
```diff
@@ -74,3 +91,17 @@ VALUES ('sliquff9-4oki-zu9u-tlw6-fb0lj3q5r7wh', 'false-email-address@mail.com',
...
+      ( 'Label 2'),
+      ( 'Label 3'),
+      ( 'Label 4'),
+      ( 'Label 5'),
+      ( 'Label 6'),
+      ( 'Label 7'),
+      ( 'Label 8'),
+      ( 'Label 9'),
+      ( 'Label 10'),
+      ( 'Label 11');
```
Comment (jpastena) [inline database/createDB.sql]: new line please

File: server/src/main/java/com/example/domain/Label.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+@NoArgsConstructor
+@Builder
+public class Label {
+
+	@Id
+	@NotEmpty
+	@Length(max = 50)
+	private String name;
+
+	public Label(String name) {
```
Comment (jpastena) [inline server/src/main/java/com/example/domain/Label.java]: if you are using `@AllArgsConstructor` you don't need this.

File: server/src/main/java/com/example/domain/mapper/LabelMapper.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+import com.example.domain.Label;
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Component;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Component
+@AllArgsConstructor
+public class LabelMapper {
```
Comment (jpastena) [inline server/src/main/java/com/example/domain/mapper/LabelMapper.java]: Do you need this class?

File: server/src/main/java/com/example/dto/mapper/LabelDtoMapper.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package com.example.dto.mapper;
+
+import com.example.domain.Label;
+import com.example.dto.LabelDto;
+import org.springframework.stereotype.Component;
+
+@Component
+public class LabelDtoMapper {
```
Comment (jpastena) [inline server/src/main/java/com/example/dto/mapper/LabelDtoMapper.java]: This is okay, but I wanted to give you a bit more insight about DTO and Entity.  
The entity is a 1:1 mapping of a java object with a table in the database.  
A dto, is a json/java object that is used to transfer data between frontend and backend.  
it's possible to use an entity as dto.  
In our application, we wanted to have different "views" of the same entity Profile, sometimes we want to see profile with machines, sometimes without. we can do that by using the same Entity and mapping to different DTOs.  
In case of the Label, the label-dto and the label-entity is the same, so the whole label-dto add a bit of overheard, without accomplishing much. but is fine. just wanted to point out for the next time you are in a situation of "I need to create a DTO for this entity?"

File: server/src/main/java/com/example/service/ProfileService.java [hunk 11]
```diff
@@ -38,12 +41,17 @@ public Profile addOne(NewProfileDto newProfileDto) {
 		                                       .map(Machine::getAssetName)
 		                                       .collect(Collectors.toList());
 		List<Machine> machines = this.machineRepository.findAllById(machineIds);
+		List<String> labelIds = newProfileDto.getLabels()
+				.stream()
+				.map(Label::getName)
+				.collect(Collectors.toList());
+		List<Label> labels = this.labelRepository.findAllById(labelIds);
```
Comment (jpastena) [inline server/src/main/java/com/example/service/ProfileService.java]: Can you add a check on the labels.size and labelIds size, like is done for the machines? (see the next `if` block).  
this prevent that the user tries to insert labels that are not present in the database, making the following `save` function to crash.

File: server/src/main/java/com/example/service/ProfileService.java [hunk 12]
```diff
@@ -58,7 +66,8 @@ public ProfileDto save(String id, EditProfile updateProfile) {
 		Profile profile = findProfileById(id);
 
 		profile.setName(updateProfile.getName());
-		profile.setDescription(updateProfile.getDescription());
+        profile.setDescription(updateProfile.getDescription());
+        profile.setLabels(updateProfile.getLabels());
```
Comment (jpastena) [inline server/src/main/java/com/example/service/ProfileService.java]: check that labels in updateProfile, are in the database. (see the code in the `addOne` function as reference)

File: server/src/main/java/com/example/domain/Label.java (hunk 8 above)
Comment (eileenabadi) [reply server/src/main/java/com/example/domain/Label.java]: If I delete the constructor, a function in Label Services is error, because it needs it.
If I delete @Allargs... The set profile on this file is error, 
If I keep both, it's fine.

File: server/src/test/java/com/example/demo/integration/ProfileControllerTest.java [hunk 13]
```diff
@@ -62,6 +63,7 @@
 	private static final String MACHINE_0_ID = "PN0000";
 	private static final String MACHINE_1_ID = "PN0001";
 	private static final String MACHINE_2_ID = "PN0002";
+	private static final String LABEL_0_ID = "Label 1";
```
[the remark this answers is not in the document]
Comment (eileenabadi) [reply server/src/test/java/com/example/demo/integration/ProfileControllerTest.java]: I don't get any error, so i think it's fine

File: server/src/main/java/com/example/service/ProfileService.java (hunk 12 above)
Comment (eileenabadi) [reply server/src/main/java/com/example/service/ProfileService.java]: label which shows in the list are from DB, so it should be existed in DB,
I don't get why should we create another function to check if it's existed.

File: client/src/app/main/profile/label/profile-label.component.scss [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+  padding: 20px;
+}
+
+.label-title {
+  font-weight: bold;
+  font-size: 20px;
+}
+
+.label-table {
+  margin-top: 20px;
```
Comment (mischaboldy) [inline client/src/app/main/profile/label/profile-label.component.scss]: you can just use the margin-top class i think

File: client/src/app/main/profile/new/profile-new.component.ts [hunk 15]
```diff
@@ -20,19 +23,61 @@ export class ProfileNewComponent implements OnInit {
...
   }
 
   public ngOnInit() {
     this.profile = new Profile();
+    this.dropdownSettings = this._dropdown.dropdownSettings;
     this.newProfileFormGroup = this._formBuilder.group(NewProfileFormFields);
     this.newProfileFormGroup.reset();
     this.newProfileFormGroup.get('machines').setValue([]);
+
+    this._machineService.find()
```
Comment (mischaboldy) [inline client/src/app/main/profile/new/profile-new.component.ts]: I think this is moved in another PR to a separate retrieveMachines function (because of the filtering), not sure if you still need this

File: database/createDB.sql [hunk 16]
```diff
@@ -74,3 +91,17 @@ VALUES ('sliquff9-4oki-zu9u-tlw6-fb0lj3q5r7wh', 'false-email-address@mail.com',
 	   	 ('3a4l001f-l988-dt8r-6i9q-whs2u597t3pz', 'email-address@mail.com', 'password', 0),
 	   	 ('d0hevl5o-zldd-p9j5-v6p0-b59un7v41er3', 'true-email-address@mail.com', 'password', 1),
 	     ('yk34mf7q-xf42-vq5g-cs3o-rix6g0b2upyx', 'email-address@mail.com', 'password', 0);
+
+--seed LABEL table
+INSERT INTO LABEL (NAME)
+VALUES( 'Label 1'),
```
Comment (mischaboldy) [inline database/createDB.sql]: do we really need to seed the DB with labels?

Comment (eileenabadi) [reply database/createDB.sql]: i think we do, if you want a list of available labels as an options when adding profiles.
it's also useful if you want to categorize profile by label
